rsocket-core/src/main/java/io/rsocket/metadata/CompositeMetadataFlyweight.java

```diff
@@ -0,0 +1,492 @@
+ * composite metadata entry can be also manipulated as an higher abstraction {@link Entry} class,
+ * which provides its own encoding and decoding primitives.
+ */
+public class CompositeMetadataFlyweight {
+
+  static final int STREAM_METADATA_KNOWN_MASK = 0x80; // 1000 0000
+  static final byte STREAM_METADATA_LENGTH_MASK = 0x7F; // 0111 1111
+
+  /**
+   * Denotes that an attempt at 0-garbage decoding failed because the input buffer didn't have
+   * enough bytes to represent a complete metadata entry, only part of the bytes.
+   */
+  public static final ByteBuf[] METADATA_MALFORMED = new ByteBuf[0];
+  /**
+   * Denotes that an attempt at garbage-free decoding failed because the input buffer was completely
+   * empty, which generally means that no more entries are present in the buffer.
+   */
+  public static final ByteBuf[] METADATA_BUFFERS_DONE = new ByteBuf[0];
+  /**
+   * Denotes that an attempt at higher level decoding of an entry components failed because the
+   * input buffer was completely empty, which generally means that no more entries are present in
+   * the buffer.
+   */
+  static final Object[] METADATA_ENTRIES_DONE = new Object[0];
+
+  private CompositeMetadataFlyweight() {}
+
+  /**
+   * Decode the next metadata entry (a mime header + content pair of {@link ByteBuf}) from a {@link
+   * ByteBuf} that contains at least enough bytes for one more such entry. The header buffer is
+   * either:
+   *
+   * <ul>
+   *   <li>made up of a single byte: this represents an encoded mime id, which can be further
+   *       decoded using {@link #decodeMimeIdFromMimeBuffer(ByteBuf)}
+   *   <li>made up of 2 or more bytes: this represents an encoded mime String + its length, which
+   *       can be further decoded using {@link #decodeMimeTypeFromMimeBuffer(ByteBuf)}. Note the
+   *       encoded length, in the first byte, is skipped by this decoding method because the
+   *       remaining length of the buffer is that of the mime string.
+   * </ul>
+   *
+   * Moreover, if the source buffer is empty of readable bytes it is assumed that the composite has
+   * been decoded entirely and the {@link #METADATA_BUFFERS_DONE} constant is returned. If the
+   * buffer contains <i>some</i> readable bytes but not enough for a correct representation of an
+   * entry, the {@link #METADATA_MALFORMED} constant is returned.
+   *
+   * @param compositeMetadata the source {@link ByteBuf} that originally contains one or more
+   *     metadata entries
+   * @param retainSlices should produced metadata entry buffers {@link ByteBuf#slice() slices} be
+   *     {@link ByteBuf#retainedSlice() retained}?
+   * @return a {@link ByteBuf} slice array of length 2 containing the mime header buffer and the
+   *     content buffer, or one of the zero-length error constant arrays
+   */
+  public static ByteBuf[] decodeMimeAndContentBuffers(
+      ByteBuf compositeMetadata, boolean retainSlices) {
```

Comment:
Where is the flyweight for the items in this array?

Reply:
only the first of the two buffers needs to be decoded (IF it exists, as the array can be empty in case of malformed composite), and the methods to do so are in the same `CompositeMetadataFlyweight`:
 - `decodeMimeIdFromMimeBuffer` in case `readableBytes() == 1`
 - `decodeMimeTypeFromMimeBuffer` otherwise

Reply:
The way I was thinking about doing this was Iterator that kept track of indexes and then a flyweight you would place over the ByteBuf slices the iterator emitted. You'd have an EntryFlyweight with a couple static methods.

Reply:
Ok I see - the CompositeMetadata is the flyweight. How do you tell the demaraction between CompositeMetadata? Like if you have 8 metadatas and you want to iterator over them?

Reply:
that's where the new CompositeMetadata comes in, with hasNext / decodeNext. alternatively the flyweight method returns an empty array when all was decoded

Reply:
ok so you mutate the readerindex. That doesn't work like any of the other frames/payloads in RSocket. They all reset there read indexes are finished- they don't change the state of the ByteBuf to back where it was. We've found it's alot easier to deal with ByteBufs that way, and update the read indexes as a side-effect makes the code harder to debug.

Reply:
does that apply to slices? `decodeMime[Type|Id]FromMimeBuffer` both work from a slice of the original whole metadata buffer, so that `skipBytes(1)` in decodeMimeTypeFromMimeBuffer doesn't move the readerIndex of the whole composite metadata...

That said, the main decoding method does indeed move the full metadata buffer's `readerIndex()`. I thought it was unavoidable since the approach was to decode a subset of the whole buffer, without state keeping, and the length of the slices (metaheader and metacontent) can't be known without decoding.

But maybe that state can be externalized, as a `startFromIndex` parameter to `decodeMimeAndContentBuffers`? It would be a burden put on the user to keep track of the next `startFromIndex` to read from, though... She would have to do something like `index = index + slices[0].readableBytes() + slices[1].readableBytes() + 1`...

This can of course be done inside the `CompositeMetadata` "iterator", but for low level decoding it can be more work.

Reply:
note The best way to compute the next entry index is to provide a flyweight util method, since the decoded slices are not necessarily covering contiguous portions of the orignal full metadata buffer (namely currently they don't include the 3 bytes content length)